vowpalwabbit/shared_feature_merger.cc

```diff
@@ -0,0 +1,112 @@
+#include "shared_feature_merger.h"
+#include "cb.h"
+#include "example.h"
+#include "label_dictionary.h"
+#include "learner.h"
+#include "options.h"
+#include "parse_args.h"
+#include "vw.h"
+
+#include <iterator>
+
+namespace VW
+{
+namespace shared_feature_merger
+{
+static const std::vector<std::string> option_strings = {
+    "csoaa_ldf", "wap_ldf", "cb_adf", "explore_eval", "cbify_ldf", "cb_explore_adf", "warm_cb"};
+
+bool use_reduction(config::options_i& options)
+{
+  for (const auto& opt : option_strings)
+  {
+    if (options.was_supplied(opt))
+      return true;
+  }
+  return false;
+}
+
+struct sfm_data
+{
+  LEARNER::multi_learner* base;
+};
+
+template <bool is_learn>
+void predict_or_learn(sfm_data& data, LEARNER::multi_learner& base, multi_ex& ec_seq)
+{
+  if (ec_seq.size() == 0)
+    THROW("cb_adf: At least one action must be provided for an example to be valid.");
+  multi_ex tmp;
+  multi_ex* source = &ec_seq;
+
+  const bool has_example_header = CB::ec_is_example_header(*ec_seq[0]);
+  if (has_example_header)
+  {
+    source = &tmp;
+    // copy vector and pass it down instead
+    tmp.reserve(ec_seq.size() - 1);
+    // skip first element since its a shared feature
+    auto startIter = std::next(ec_seq.begin());
+    tmp.insert(tmp.end(), startIter, ec_seq.end());
+    // merge sequences
+    for (size_t k = 0; k < tmp.size(); k++) LabelDict::add_example_namespaces_from_example(*tmp[k], *ec_seq[0]);
+    tmp[0]->pred = ec_seq[0]->pred;
+  }
+  if (source->size() == 0)
+    return;
+  if (is_learn)
+    base.learn(*source);
```

Comment:
Instead of dealing with source, why not just pass ec_seq?

Reply:
This is one of two ways I can think of to handle this operation. The alternative would be to remove the shared example from the vector before passing it down to the rest of the stack.

Removing the shared example has the benefit of being a little faster and not necessarily requiring any memory allocations, the gains are fairly minimal in this case since the vector only maintain pointers to examples (and removing the first element of a vector requires a memcpy as well)
I had originally thought of changing multi_ex to be a deque instead, but Jack pointed out that the object is exposed in the various interfaces, which would complicate that change

Reply:
Let's avoid memory allocation.

There is a third approach which is a bit unusual----you adjust pointers directly.

Reply:
re-arranging the pointers should have a similar (or worse) performance to just erasing and re-inserting the first element. Looking at the docs, erase has some guarantees around iterator invalidation, which means memory will never be reallocated for an erase; in turn this shouldn't affect the capacity at all, which would mean the insertion shouldn't cause a realloc either.